woocommerce/woocommerce
# Add slotFills for header items

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds slotFills to the product editor header.

<img width="529" alt="Screenshot 2024-01-16 at 15 21 44" src="https://github.com/woocommerce/woocommerce/assets/1314156/1beaa5b3-21b7-445f-b78b-1403e6781ee3">

### Extensibility

**PluginHeaderItemPopover**
```
<PluginHeaderItemPopover
    label={}
    icon={}
>
    <div>Popover content</div>
</PluginHeaderItemPopover>
```

**PluginHeaderItemModal**

```
<PluginHeaderItemModal
    label={}
    icon={}
    title={ 'My title' }
>
  <div>Modal content</div>
</PluginHeaderItemModal>
```

These SlotFills allow for the header to be extended by extensions with popovers and modals. We are using `PinnedItems.Slot` under the hood.

This PR does what the issue requests, but [I left in the issue a few extra questions](https://github.com/woocommerce/woocommerce/issues/42805#issuecomment-1893296138) regarding a follow-up for this functionality. 

@mattsherman, in the issue was suggested to use `PinnedItems`, but [it doesn't order the `fills`](https://github.com/WordPress/gutenberg/blob/trunk/packages/interface/src/components/pinned-items/index.js) items. This is why I removed the prop from the components. 

Closes #42805.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure the feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`.
2. To make testing easier, I created [this plugin](https://github.com/woocommerce/woocommerce/files/13996806/myplugin.zip). Please download it, install and activate it.
3. Go to `Products > Add New`.
4. Verify the items were added to the product header. 

![Screen Capture on 2024-01-16 at 15-35-27](https://github.com/woocommerce/woocommerce/assets/1314156/876d49e9-a080-44df-9b6f-23adc4db2051)

5. Click them and verify they work as expected.

The testing plugin will add the following components:
```
<PluginHeaderItemPopover
	icon={ plugins }
	label={ 'My plugin' }
>
	<div className="my-plugin-class-name">
		<p>
			Popover without button or callback.
		</p>
	</div>
</PluginHeaderItemPopover>
<PluginHeaderItemPopover
	icon={ comment }
	label={ 'Comment extension' }
>
	{ ( { isVisible, setVisible }: { isVisible: boolean, setVisible: ( visible: boolean ) => void } ) => (
		<div className="my-plugin-class-name">
			<p>
				Popover with close button.
			</p>
			<Button
				onClick={ ()=> {
					console.log( `The popover is: ${ isVisible ? 'visible' : 'hidden' }, let's hide it!`)
					setVisible( false );
				} }
				variant="primary"
			>
				Close popover!
			</Button>
		</div>
	) }
</PluginHeaderItemPopover>
<PluginHeaderItemModal
	icon={ arrowDown }
	label={ 'Modal label' }
	title={ 'Title' }
>
	{ ( { isOpen, setOpen } ) => (
		<div className="my-plugin-class-name">
			<p>
				Modal with close button.
			</p>
			<Button
				onClick={ ()=> {
					console.log( `The modal is: ${ isOpen ? 'open' : 'hidden' }, let's close it!`)
					setOpen( false );
				} }
				variant="primary"
			>
				Close modal!
			</Button>
		</div>
	) }
</PluginHeaderItemModal>
<PluginHeaderItemModal
	icon={ info }
	label={ 'Modal label' }
	title={ 'Title' }
>
	<div className="my-plugin-class-name">
		<p>
			Modal without button or callback.
		</p>
	</div>
</PluginHeaderItemModal>
```



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
